---
title: "coproID run report"
output: html_document
---

# coproID report

Report generated on the `r Sys.Date()`

![](coproID_nf-core_logo_small.png)

## Introduction
[coproID](https://github.com/nf-core/coproID) is a pipeline to identify the source of coprolites, and in general, of a metagenomic sample.

If you read these lines, coproID successfully finished running and you can find your results below.  
You can find more informations about the different result files in the coproID documentation: [coproid.readthedocs.io/](https://coproid.readthedocs.io/en/latest/output.html)


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
require(ggplot2)
require(rmarkdown)
require(magrittr)
require(DT)
require(plotly)
require(gridExtra)
```


```{r, echo=FALSE, include=FALSE}
d = read.csv('coproID_result.csv', row.names = 1)
```

## coproID summary table

```{r, table1, echo=FALSE}
# d %>%
#   knitr::kable(format = "html", col.names = colnames(d)) %>%
#   kableExtra::kable_styling() %>%
#   kableExtra::scroll_box(width = "100%", height = "400px")
d %>% 
  datatable(
    extensions = 'Buttons',
    width = '80%',
    options = list(dom = 'Bfrtip', 
                   buttons = c('excel', "csv"),
                   autowidth='True'),
    caption='coproID summary table')
```


## Microbiome composition embedding


```{r, echo=FALSE}
e = read.csv("sourcepredict_embedding.csv", row.names = 1)
e['ml'] = as.factor(ifelse(e["labels"] == 'sink', "reference", "test"))


g = ggplot(data=e, mapping = aes(x=PC1, y=PC2, label=name)) + geom_point(aes(color=labels, shape=ml)) + scale_color_discrete(name='Organism') + scale_shape_discrete(name='Reference') + theme_classic() + labs(x='DIM1', y='DIM2')
ggplotly(g)
```

## Damage profiles

```{r, echo=FALSE, results='asis'}
files <- list.files(pattern = "\\_freq.txt$")

samp_names = c()
for (i in seq(1, length(files))){
  # print(files[i])
  afile = files[i]
  spt = strsplit(as.character(afile), "[.]")
  samp_name = spt[[1]][1]
  samp_names = append(samp_names, samp_name)
}
samp_names = unique(samp_names)

for (i in seq(1, length(samp_names))){
  samp_name = samp_names[i]
  print(samp_name)
  cat('\n')  
  cat("### Sample: ", strsplit(samp_name, "_otu_")[[1]][1]," - species: ", strsplit(samp_name, "_otu_")[[1]][2], "\n")
  fwd = paste(append(samp_name, "5pCtoT_freq.txt"), collapse = ".")
  fwd = read.csv(fwd, skip=3, sep="\t", col.names = c('pos','X5pCtoT'))
  rev = paste(append(samp_name, "3pGtoA_freq.txt"), collapse = ".")
  rever = read.csv(rev, skip=3, sep="\t", col.names = c('pos','X3pGtoA'))
  rever$pos = rev(rever$pos * -1)
  rever$X3pGtoA = rev(rever$X3pGtoA)
  
  f = ggplot(fwd, aes(x=pos,y=X5pCtoT)) + geom_line() + labs(title='5pC>t', y="")
  r = ggplot(rever, aes(x=pos,y=X3pGtoA)) + geom_line() + labs(title='3pG>A', y = "")
  grid.arrange(f, r, nrow = 1)
  cat('\n')  
}

```

