# Output

## coproID results

-   `results/coproID_result.html`  

     This file is the report of **coproID** and contains explanations on how the calculation were performed, the files used, and your results.

     An example **coproID** output file file can be found [here](_static/coproID_result.html)

-   `results/coproid_result.csv`  

     This file contains the numerical results of **coproID**

     An example **coproID** output file file can be found [here](_static/coproid_result.csv)

-   `results/multiqc_report.html`

      This file is generated by [MultiQC](http://multiqc.info/) and reports the metrics of the different tools used by **coproID**

      An example **coproID** MultiQC report can be found [here](_static/multiqc_report.html)

-   `results/mapdamage_{target_organism}/*`

      [MapDamage ](https://ginolhac.github.io/mapDamage/) plots. These plots are included in the **coproID** reports.

-   `results/bam/{target_organism}/*bam`

      The filtered alignment `bam` files on each target genome

## coproID logs

You can generate a log file report of coproID run by using nextflow reporting option `-with-report`

Example:

    nextflow run maxibor/coproid --genome1 'path/to/data/genomes/hsapiens.fa' --genome2 'path/to/data/genomes/cfamiliaris.fa' --name1 'data/genomes/Homo_sapiens' --name2 'Canis_familiaris' --reads '*_R{1,2}.fastq.gz' -with-report report.html

This log file will be saved as `report.html`
